### PR TITLE
pulseaudio: add libintl dependency on darwin

### PR DESCRIPTION
**Pull request by otavio**
Log of failure: https://hydra.nixos.org/build/174832426/nixlog/4
Refs: ZHF #172160
Signed-off-by: Otavio Salvador <otavio@ossystems.com.br>

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by siraben** [comment]
Result of `nixpkgs-review pr 172230` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>7 packages marked as broken and skipped:</summary>
  <ul>
    <li>anki</li>
    <li>imgbrd-grabber</li>
    <li>mnemosyne</li>
    <li>projectm</li>
    <li>retro-gtk</li>
    <li>rootbar</li>
    <li>shattered-pixel-dungeon</li>
  </ul>
</details>
<details>
  <summary>11 packages failed to build:</summary>
  <ul>
    <li>castty</li>
    <li>gnomeExtensions.sound-output-device-chooser</li>
    <li>libcanberra</li>
    <li>libcanberra-gtk2</li>
    <li>libcanberra-gtk3</li>
    <li>libpulseaudio</li>
    <li>libsoundio</li>
    <li>moc</li>
    <li>pulsemixer</li>
    <li>rhvoice</li>
    <li>roc-toolkit</li>
  </ul>
</details>

**Comment by siraben** [comment]
❯ nix log /nix/store/jk3ya7d6i0nycwfmd4mrcnxxz2zsaanq-libpulseaudio-15.0.drv | ix
http://ix.io/3XoZ

**Comment by otavio** [comment]
@siraben I am on Linux so it is hard for me to debug this. Could you take  a look?

**Comment by tjni** [comment]
When I was looking at this earlier, I found the same issue building locally. After passing the `libintl` issue, the `dbus` issue appeared. Passing the `dbus` issue, other ones appeared. In my opinion, we should merge this since it seems necessary and does not make things worse, even though it does not fix the package on Darwin.

**Comment by tjni** [comment]
Note that `libintl` is defined as:

```
# On non-GNU systems we need GNU Gettext for libintl.
libintl = if stdenv.hostPlatform.libc != "glibc" then gettext else null;
```

which means, depending on taste, it might be preferred to include `libintl` on all systems, not only Darwin.

**Comment by thiagokokada** [comment]
> When I was looking at this earlier, I found the same issue building locally. After passing the `libintl` issue, the `dbus` issue appeared. Passing the `dbus` issue, other ones appeared. In my opinion, we should merge this since it seems necessary and does not make things worse, even though it does not fix the package on Darwin.

One thing that I found strange is that `dbus` shouldn't be used in macOS AFAIK (it is a Linux specific thing).

I am trying to do an investigation in my local machine right now. Don't think it is worth merging it in the current state.

**Comment by tjni** [comment]
> One thing that I found strange is that `dbus` shouldn't be used in macOS AFAIK (it is a Linux specific thing).
> 
> I am trying to do an investigation in my local machine right now. Don't think it is worth merging it in the current state.

Sounds good! Let me know if I can help. I (only) have an M1 machine.

**Comment by thiagokokada** [comment]
> > One thing that I found strange is that `dbus` shouldn't be used in macOS AFAIK (it is a Linux specific thing).
> > I am trying to do an investigation in my local machine right now. Don't think it is worth merging it in the current state.
> 
> Sounds good! Let me know if I can help. I (only) have an M1 machine.

Now it starts to build, but it fails during compilation. I think this may be related to the compiler maybe, or something else.

**Comment by thiagokokada** [comment]
Now maybe it makes sense to merge as-is since at least it tries to build. Still not sure if we want it or not.

BTW, from the looks of it, this release (15.0) seems problematic on Darwin. Both macports and homebrew are still packaging the 14.2 release.

**Comment by tjni** [comment]
If `dbus` should not be used on OSX, it seems we need to disable it explicitly. In `meson_options.txt`, the `dbus` feature is declared `auto`, and `nixpkgs` defaults `auto` to `required`. We could fix this by adding to `mesonFlags`:

```
++ lib.optional stdenv.isDarwin "-Ddbus=disabled"
```

You're right that this then runs into further issues, and I agree it could be best to take it one step at a time.

**Comment by tjni** [comment]
I also get compile-time errors, such as:

```
In file included from ../src/pulsecore/cpu-x86.c:28:
/nix/store/qlicbgwj6mddbcdcfsqqp7g5fq3rphd1-clang-wrapper-11.1.0/resource-root/include/cpuid.h:11:2: error: this header is for x86 only
#error this header is for x86 only
```

which looks related to being on an M1 Mac.

If we want to try to continue with this version, I would advocate for committing the following:

1. Add `libintl` as a build time and run time dependency (I like unconditionally, but it can also be conditional on Darwin).
2. Disable `oss-output` and `dbus` on Darwin (at least for now)
3. Add `glib` as a build time and run time dependency (so that `gio-2` is available)

That seems to get us past the configuration phase, and then we can try to tackle the compilation (or mark as broken if we can't figure it out).

I suppose we could also continue using 14.2 on Darwin?

File: pkgs/servers/pulseaudio/default.nix [hunk 1]
```diff
@@ -105,8 +103,10 @@ stdenv.mkDerivation rec {
     "-Dsysconfdir_install=${placeholder "out"}/etc"
     "-Dudevrulesdir=${placeholder "out"}/lib/udev/rules.d"
   ]
-    ++ lib.optional (stdenv.isLinux && useSystemd) "-Dsystemduserunitdir=${placeholder "out"}/lib/systemd/user"
-    ;
+  ++ lib.optional (stdenv.isLinux && useSystemd) "-Dsystemduserunitdir=${placeholder "out"}/lib/systemd/user"
+  ++ lib.optionals (stdenv.isDarwin) [
+    "-Doss-output=disabled"
+  ];
```
Comment (otavio) [inline pkgs/servers/pulseaudio/default.nix:109]:
```suggestion
  ++ lib.optional (stdenv.isDarwin) "-Doss-output=disabled";
```

**Comment by otavio** [comment]
@tjni keep the last known working version does make sense.

**Comment by thefloweringash** [comment]
I’ve also spent some time on pulseaudio on Darwin but I don’t think I have the time to see it through at the moment. Here’s where I got to: https://github.com/NixOS/nixpkgs/compare/master...thefloweringash:libpulseaudio-darwin?expand=1

**Comment by thiagokokada** [comment]
> > If dbus should not be used on OSX, it seems we need to disable it explicitly.
> 
> Ignore what I said before. Dbus is used both for Homebrew and MacPorts.
> 
> I still think this is mostly unnecessary for macOS, but if it builds and works we should keep it.

https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/pulseaudio.rb#L55
https://github.com/macports/macports-ports/blob/master/audio/pulseaudio/Portfile#L49

Sorry, so dbus is disabled in Homebrew and enabled in MacPorts. I think we are not really interested in `dbus` inside Darwin (it does seems to build at least though), so yeah, maybe it makes sense to disable it.

**Comment by thiagokokada** [comment]
> I’ve also spent some time on pulseaudio on Darwin but I don’t think I have the time to see it through at the moment. Here’s where I got to: https://github.com/NixOS/nixpkgs/compare/master...thefloweringash:libpulseaudio-darwin?expand=1

Nice. I took your PR and did a clean-up and made it build: https://github.com/NixOS/nixpkgs/pull/172307

**Pull request closed without merging**